File: app/src/main/java/org/pragyan/dalal18/fragment/MortgageFragment.java
package org.pragyan.dalal18.fragment;


import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.loader.app.LoaderManager;
import androidx.loader.content.Loader;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.appcompat.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.NumberPicker;
import android.widget.RadioGroup;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import org.pragyan.dalal18.R;
import org.pragyan.dalal18.dagger.ContextModule;
import org.pragyan.dalal18.dagger.DaggerDalalStreetApplicationComponent;
import org.pragyan.dalal18.ui.MainActivity;
import org.pragyan.dalal18.utils.ConnectionUtils;
import org.pragyan.dalal18.utils.Constants;
import org.pragyan.dalal18.utils.StockUtils;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import dalalstreet.api.DalalActionServiceGrpc;
import dalalstreet.api.actions.GetMortgageDetailsResponse;
import dalalstreet.api.actions.MortgageStocksRequest;
import dalalstreet.api.actions.MortgageStocksResponse;
import dalalstreet.api.actions.RetrieveMortgageStocksRequest;
import dalalstreet.api.actions.RetrieveMortgageStocksResponse;

import static org.pragyan.dalal18.utils.StockUtils.getCompanyNameFromStockId;
import static org.pragyan.dalal18.utils.StockUtils.getQuantityOwnedFromCompanyName;

/* Uses GetMortgageDetails() for setting stocksMortgaged (int data member)
*  Uses MortgageStocks() to mortgage stocks
*  Uses RetrieveStocksFromMortgage() to get back mortgaged stocks */
public class MortgageFragment extends Fragment implements LoaderManager.LoaderCallbacks<GetMortgageDetailsResponse>{

    @Inject
    DalalActionServiceGrpc.DalalActionServiceBlockingStub actionServiceBlockingStub;

    @BindView(R.id.mortgageRadioGroup)
    RadioGroup mortgageRadioGroup;

    @BindView(R.id.stocks_editText)
    EditText stocksEditText;

    @BindView(R.id.stocksOwned_textView)
    TextView ownedTextView;

    @BindView(R.id.stocksMortgaged_textView)
    TextView mortgagedTextView;

    @BindView(R.id.currentPrice_textView)
    TextView currentPriceTextView;

    @BindView(R.id.mortgageRate_textView)
    TextView mortgageRateTextView;

    @BindView(R.id.mortgageRateText_textView)
    TextView mortgageRateTextTextView;

    @BindView(R.id.depositPrice_textView)
    TextView depositPriceTextView;

    @BindView(R.id.depositPriceText_textView)
    TextView depositPriceTextTextView;

    @BindView(R.id.mortgage_button)
    Button mortgageButton;

    int stocksOwned = 0, stocksMortgaged = 0, stocksTransaction, lastStockId = 1;
    Spinner companySpinner;
    String [] companiesArray;

    private ConnectionUtils.OnNetworkDownHandler networkDownHandler;
    private AlertDialog loadingDialog;

    public MortgageFragment() {
        // Required empty public constructor
    }

    private BroadcastReceiver refreshStockPricesReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (getActivity() != null && intent.getAction() != null && intent.getAction().equalsIgnoreCase(Constants.REFRESH_STOCK_PRICES_ACTION)) {
                getActivity().getSupportLoaderManager().restartLoader(Constants.MORTGAGE_DETAILS_LOADER_ID, null, MortgageFragment.this);
            }
        }
    };

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            networkDownHandler = (ConnectionUtils.OnNetworkDownHandler) context;
        } catch (ClassCastException classCastException) {
            throw new ClassCastException(context.toString() + " must implement network down handler.");
        }
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (getContext() != null){
            View dialogView = LayoutInflater.from(getContext()).inflate(R.layout.progress_dialog, null);
            String tempString = "Getting mortgage details...";
            ((TextView) dialogView.findViewById(R.id.progressDialog_textView)).setText(tempString);
            loadingDialog = new AlertDialog.Builder(getContext())
                    .setView(dialogView)
                    .setCancelable(false)
                    .create();
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        final View rootView = inflater.inflate(R.layout.fragment_mortgage, container, false);
        DaggerDalalStreetApplicationComponent.builder().contextModule(new ContextModule(getContext()) ).build().inject(this);
        ButterKnife.bind(this, rootView);
        if (getActivity() != null) getActivity().setTitle("Mortgage Stocks");

        companiesArray = StockUtils.getCompanyNamesArray();

        companySpinner = rootView.findViewById(R.id.mortgage_companies_spinner);
        ArrayAdapter<String> arrayAdapter = new ArrayAdapter<>(getActivity(), R.layout.company_spinner_item, companiesArray);
        companySpinner.setAdapter(arrayAdapter);

        companySpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                getActivity().getSupportLoaderManager().restartLoader(Constants.MORTGAGE_DETAILS_LOADER_ID, null, MortgageFragment.this);
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });

        mortgageRadioGroup.setOnCheckedChangeListener((radioGroup, id) -> {
            int currentPrice = StockUtils.getPriceFromStockId(MainActivity.Companion.getGlobalStockDetails(), lastStockId);

            if (id == R.id.mortgage_radioButton) {
                mortgageButton.setText(R.string.mortgage_uppercase);
                mortgageRateTextTextView.setText(R.string.mortgage_deposit_rate);

                String temp = " :  " + String.valueOf(Constants.MORTGAGE_DEPOSIT_RATE) + "%";
                mortgageRateTextView.setText(temp);

                depositPriceTextTextView.setText(R.string.deposit_rate_per_stock);

                temp = " :  " + Constants.RUPEE_SYMBOL + " " + String.valueOf(Constants.MORTGAGE_DEPOSIT_RATE * currentPrice / 100);
                depositPriceTextView.setText(temp);

            } else {
                mortgageButton.setText(R.string.retrieve_uppercase);
                mortgageRateTextTextView.setText(R.string.mortgage_retrival_rate);

                depositPriceTextTextView.setText(R.string.retrieval_rate_per_stock);

                String temp = " :  " + String.valueOf(Constants.MORTGAGE_RETRIEVE_RATE) + "%";
                mortgageRateTextView.setText(temp);

                temp = " :  " + Constants.RUPEE_SYMBOL + " " + String.valueOf(Constants.MORTGAGE_RETRIEVE_RATE * currentPrice / 100);
                depositPriceTextView.setText(temp);
            }
        });

        String ownedString = "N/A";
        ownedTextView.setText(ownedString);

        String mortgageString = "N/A";
        mortgagedTextView.setText(mortgageString);

        return rootView;
    }

    @OnClick(R.id.mortgage_button)
    void onMortgageButtonClick() {

        if (stocksEditText.getText().toString().trim().isEmpty()) {
            stocksEditText.setError("Stocks quantity missing");
            stocksEditText.requestFocus();
            return;
        } else {
            stocksEditText.setError(null);
            stocksEditText.clearFocus();
        }

        stocksTransaction = Integer.parseInt(stocksEditText.getText().toString().trim());

        if (mortgageRadioGroup.getCheckedRadioButtonId() == R.id.mortgage_radioButton) {

            if (stocksTransaction <= stocksOwned) {
                MortgageStocksResponse mortgageStocksResponse = actionServiceBlockingStub.mortgageStocks(
                        MortgageStocksRequest
                                .newBuilder()
                                .setStockId(companySpinner.getSelectedItemPosition()+1)
                                .setStockQuantity(Integer.parseInt(stocksEditText.getText().toString()))
                                .build()
                );

                if (mortgageStocksResponse.getStatusCode().getNumber() == 0) {
                    Toast.makeText(getContext(), "Transaction successful", Toast.LENGTH_SHORT).show();

                    stocksOwned -= stocksTransaction;
                    stocksMortgaged += stocksTransaction;

                    String ownedString = " :  " + String.valueOf(stocksOwned);
                    ownedTextView.setText(ownedString);

                    String mortgageString = " :  " + String.valueOf(stocksMortgaged);
                    mortgagedTextView.setText(mortgageString);

                    stocksEditText.setText("");

                } else {
                    Toast.makeText(getContext(), mortgageStocksResponse.getStatusMessage(), Toast.LENGTH_SHORT).show();
                }

            } else {
                Toast.makeText(getActivity(), "Insufficient Stocks", Toast.LENGTH_SHORT).show();
            }

        } else if (mortgageRadioGroup.getCheckedRadioButtonId() == R.id.retrieve_radioButton) {

            if (stocksTransaction <= stocksMortgaged && stocksMortgaged >= 0) {
                RetrieveMortgageStocksResponse retrieveStocksResponse = actionServiceBlockingStub.retrieveMortgageStocks(
                        RetrieveMortgageStocksRequest
                                .newBuilder()
                                .setStockId(companySpinner.getSelectedItemPosition()+1)
                                .setStockQuantity(Integer.parseInt(stocksEditText.getText().toString()))
                                .build()
                );

                if (retrieveStocksResponse.getStatusCode().getNumber() == 0) {
                    Toast.makeText(getContext(), "Transaction successful", Toast.LENGTH_SHORT).show();

                    stocksOwned += stocksTransaction;
                    stocksMortgaged -= stocksTransaction;

                    String ownedString = " :  " + String.valueOf(stocksOwned);
                    ownedTextView.setText(ownedString);

                    String mortgageString = " :  " + String.valueOf(stocksMortgaged);
                    mortgagedTextView.setText(mortgageString);

                    stocksEditText.setText("");

                } else {
                    Toast.makeText(getContext(), retrieveStocksResponse.getStatusMessage(), Toast.LENGTH_SHORT).show();
                }

            } else {
                Toast.makeText(getActivity(), "You don't have sufficient stocks mortgaged", Toast.LENGTH_SHORT).show();
            }
        } else {
            Toast.makeText(getContext(), "Select mortgage or retrieve", Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public Loader<GetMortgageDetailsResponse> onCreateLoader(int id, Bundle args) {
        loadingDialog.show();
        companySpinner.setEnabled(false);

        if (getContext() != null) {
//            return new MortgageDetailsLoader(getContext(), actionServiceBlockingStub);
        } else {
            return null;
        }
    }

    @Override
    public void onLoadFinished(Loader<GetMortgageDetailsResponse> loader, GetMortgageDetailsResponse response) {

        loadingDialog.dismiss();

        if (response == null) {
            networkDownHandler.onNetworkDownError();
            return;
        }

        int stockId = StockUtils.getStockIdFromCompanyName(companiesArray[companySpinner.getSelectedItemPosition()]);
        companySpinner.setEnabled(true);

        if (response.getMortgageMapMap().get(stockId) != null)
            stocksMortgaged = response.getMortgageMapMap().get(stockId);
        else
            stocksMortgaged = 0;

        String mortgageString = " :  " + String.valueOf(stocksMortgaged);
        mortgagedTextView.setText(mortgageString);

        stocksOwned = getQuantityOwnedFromCompanyName(MainActivity.Companion.getOwnedStockDetails(), getCompanyNameFromStockId(stockId));

        String ownedString = " :  " + String.valueOf(stocksOwned);
        ownedTextView.setText(ownedString);

        lastStockId = stockId;
        int currentPrice = StockUtils.getPriceFromStockId(MainActivity.Companion.getGlobalStockDetails(), stockId);

        String tempString = " :  " + Constants.RUPEE_SYMBOL + " " + String.valueOf(currentPrice);
        currentPriceTextView.setText(tempString);

        if (mortgageRadioGroup.getCheckedRadioButtonId() == R.id.mortgage_radioButton) {
            mortgageButton.setText(R.string.mortgage_uppercase);
            mortgageRateTextTextView.setText(R.string.mortgage_deposit_rate);

            String temp = " :  " + String.valueOf(Constants.MORTGAGE_DEPOSIT_RATE) + "%";
            mortgageRateTextView.setText(temp);

            depositPriceTextTextView.setText(R.string.deposit_rate_per_stock);

            temp = " :  " + Constants.RUPEE_SYMBOL + " " + String.valueOf(Constants.MORTGAGE_DEPOSIT_RATE * currentPrice / 100);
            depositPriceTextView.setText(temp);

        } else if (mortgageRadioGroup.getCheckedRadioButtonId() == R.id.retrieve_radioButton){
            mortgageButton.setText(R.string.retrieve_uppercase);
            mortgageRateTextTextView.setText(R.string.mortgage_retrival_rate);

            depositPriceTextTextView.setText(R.string.retrieval_rate_per_stock);

            String temp = " :  " + String.valueOf(Constants.MORTGAGE_RETRIEVE_RATE) + "%";
            mortgageRateTextView.setText(temp);

            temp = " :  " + Constants.RUPEE_SYMBOL + " " + String.valueOf(Constants.MORTGAGE_RETRIEVE_RATE * currentPrice / 100);
            depositPriceTextView.setText(temp);
        }
    }

    @Override
    public void onLoaderReset(Loader<GetMortgageDetailsResponse> loader) {
        // Do nothing
    }

    @Override
    public void onResume() {
        super.onResume();
        if (getContext() != null) {
            LocalBroadcastManager.getInstance(getContext()).registerReceiver(
                    refreshStockPricesReceiver, new IntentFilter(Constants.REFRESH_STOCK_PRICES_ACTION)
            );
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        if (getContext() != null) {
            LocalBroadcastManager.getInstance(getContext()).unregisterReceiver(refreshStockPricesReceiver);
        }
    }
}

File: app/src/main/java/org/pragyan/dalal18/fragment/MortgageFragment.kt
package org.pragyan.dalal18.fragment


import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProviders
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import dalalstreet.api.DalalActionServiceGrpc
import dalalstreet.api.actions.*
import kotlinx.android.synthetic.main.fragment_mortgage.*
import org.pragyan.dalal18.R
import org.pragyan.dalal18.dagger.ContextModule
import org.pragyan.dalal18.dagger.DaggerDalalStreetApplicationComponent
import org.pragyan.dalal18.data.DalalViewModel
import org.pragyan.dalal18.data.MortgageDetails
import org.pragyan.dalal18.utils.ConnectionUtils
import org.pragyan.dalal18.utils.Constants
import org.pragyan.dalal18.utils.StockUtils
import org.pragyan.dalal18.utils.StockUtils.*
import javax.inject.Inject

/*  Uses GetMortgageDetails() for setting stocksMortgaged (int data member)
 *  Uses MortgageStocks() to mortgage stocks
 *  Uses RetrieveStocksFromMortgage() to get back mortgaged stocks */

class MortgageFragment : Fragment(){

    @Inject
    lateinit var actionServiceBlockingStub: DalalActionServiceGrpc.DalalActionServiceBlockingStub

    private var lastStockId = 1
    private var companiesArray = StockUtils.getCompanyNamesArray()

    private lateinit var model: DalalViewModel

    lateinit var networkDownHandler: ConnectionUtils.OnNetworkDownHandler
    private var loadingDialog: AlertDialog? = null

    private val mortgageDetailsList = mutableListOf<MortgageDetails>()

    private val refreshStockPricesReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {

            //here implement both listener
            if (intent.action != null && intent.action.equals(Constants.REFRESH_STOCK_PRICES_ACTION, ignoreCase = true)) {

                currentPrice_textView.text = model.globalStockDetails[mortgage_companies_spinner.selectedItemPosition+1].price.toString()


            }
            if(intent.action != null && intent.action.equals(Constants.REFRESH_MORTGAGE_UPDATE_ACTION,ignoreCase = true))
            {
                val quantity = intent.getLongExtra(STOCKS_QUANTITY_KEY, 0)

                val stockId = intent.getIntExtra(STOCKS_ID_KEY, 0)
                val price = intent.getLongExtra(STOCKS_PRICE_KEY, 0)

                val ownedString = " :  " + getQuantityOwnedFromStockId(model.ownedStockDetails, stockId).toString()
                stocksOwnedTextView.text = ownedString

                if (quantity < 0) /* Mortgage Action */ {

                    mortgageDetailsList.forEachIndexed { index, mortgageDetails ->
                        if (mortgageDetails.stockId == stockId && mortgageDetails.mortgagePrice == price) {
                            mortgageDetails.stockQuantity -= quantity
                            return
                        }
                    }

                    mortgageDetailsList.add(MortgageDetails(stockId, -quantity, price))

                } else /* Retrieve Action */ {
                    mortgageDetailsList.forEachIndexed { index, mortgageDetails ->
                        if (mortgageDetails.stockId == stockId && mortgageDetails.mortgagePrice == price) {
                            if (mortgageDetails.stockQuantity == quantity) {
                                mortgageDetailsList.remove(mortgageDetails)
                                return
                            } else {
                                mortgageDetails.stockQuantity -= quantity
                                return
                            }
                        }
                    }
                }
            }
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        try {
            networkDownHandler = context as ConnectionUtils.OnNetworkDownHandler
        } catch (classCastException: ClassCastException) {
            throw ClassCastException(context.toString() + " must implement network down handler.")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (context != null) {
            val dialogView = LayoutInflater.from(context).inflate(R.layout.progress_dialog, null)
            val tempString = "Getting mortgage details..."
            (dialogView.findViewById<View>(R.id.progressDialog_textView) as TextView).text = tempString
            loadingDialog = AlertDialog.Builder(context!!).setView(dialogView).setCancelable(false).create()

//            retrieveAdapter = RetrieveRecyclerAdapter(context, null, this)
        }
    }
    //todo -> ask why are we not setting the price in the current price text box of the layout
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        val rootView = inflater.inflate(R.layout.fragment_mortgage, container, false)

        model = activity?.run { ViewModelProviders.of(this).get(DalalViewModel::class.java) } ?: throw Exception("Invalid activity")
        DaggerDalalStreetApplicationComponent.builder().contextModule(ContextModule(context!!)).build().inject(this)

        return rootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        (activity as AppCompatActivity).supportActionBar?.title = resources.getString(R.string.mortgage)


        with(mortgage_companies_spinner) {
            adapter = ArrayAdapter(context!!, R.layout.company_spinner_item, companiesArray)

            onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                    val ownedString = " :  " + getQuantityOwnedFromStockId(model.ownedStockDetails, mortgage_companies_spinner.selectedItemPosition + 1).toString()
                    stocksOwnedTextView.text = ownedString
                    val mortgageString = " :  " + mortgageDetailsList[mortgage_companies_spinner.selectedItemPosition + 1].stockQuantity
                    stocksMortgagedTextView.text = mortgageString
                }

                override fun onNothingSelected(parent: AdapterView<*>) {

                }
            }
        }

        mortgage_button.setOnClickListener { onMortgageButtonClick() }

    }

    private fun onMortgageButtonClick() {

        with(stocks_editText) {
            if (text.toString().trim { it <= ' ' }.isEmpty()) {
                error = "Stocks quantity missing"
                requestFocus()
                return
            } else {
                error = null
                clearFocus()
            }
        }

        val stocksTransaction = (stocks_editText.text.toString().trim { it <= ' ' }).toLong()

        if (stocksTransaction <= getQuantityOwnedFromCompanyName(model.ownedStockDetails, getCompanyNameFromStockId(lastStockId))) {
            val mortgageStocksResponse = actionServiceBlockingStub.mortgageStocks(
                    MortgageStocksRequest.newBuilder().setStockId(mortgage_companies_spinner.selectedItemPosition + 1)
                            .setStockQuantity(stocksTransaction).build())

            if (mortgageStocksResponse.statusCode == MortgageStocksResponse.StatusCode.OK) {
                Toast.makeText(context, "Transaction successful", Toast.LENGTH_SHORT).show()
                stocks_editText.setText("")
            } else
                Toast.makeText(context, mortgageStocksResponse.statusMessage, Toast.LENGTH_SHORT).show()

        } else {
            Toast.makeText(activity, "Insufficient Stocks", Toast.LENGTH_SHORT).show()
        }
    }



    override fun onResume() {
        super.onResume()
        val intentFilter = IntentFilter(Constants.REFRESH_STOCK_PRICES_ACTION)
        intentFilter.addAction(Constants.REFRESH_MORTGAGE_UPDATE_ACTION)
        LocalBroadcastManager.getInstance(context!!).registerReceiver(refreshStockPricesReceiver, intentFilter)
    }

    override fun onPause() {
        super.onPause()
        LocalBroadcastManager.getInstance(context!!).unregisterReceiver(refreshStockPricesReceiver)
    }

    companion object {
        const val STOCKS_ID_KEY = "stocks-id-key"
        const val STOCKS_QUANTITY_KEY = "stocks-quantity-key"
        const val STOCKS_PRICE_KEY = "stocks-price-key"
    }
}